QTI SECLIB migration - all drivers disabled/stub

ldts opened the pull request.

ldts committed:
feat(qti): integrate sec_core stub

Add the sec_core driver stub and hook it into the generic QTI driver
framework. The stub remains disabled while QTISECLIB continues to
provide the active implementation.

For coverity purposes, enable the driver on the non-bootable builds for
the kodiak/rb3gen2 platform.

Change-Id: I83a0afd0a2f89d0fe31b687f4341d9bb8518ec8d
Signed-off-by: Jorge Ramirez-Ortiz <jorge.ramirez@oss.qualcomm.com>

ldts commented:
@b49020 fyi

ldts commented:
need to tidy up a bit before posting on friday- anyway, comments welcome of course

b49020 commented on drivers/qti/sec_core/sec_core.c:
```diff
@@ -0,0 +1,80 @@
+ */
+void qti_sec_core_enable_ns(void)
+{
+#if defined(HWIO_APSS_GNOC_GLADIATOR_CSR_LOCK_SECURECLR_LOW_ADDR)
```
We don't enable this macro, so can we drop this API from the initial port?

ldts replied:
yep

b49020 commented on drivers/qti/sec_core/sec_core.c:
```diff
@@ -0,0 +1,80 @@
+	mmio_write_32(HWIO_APSS_CL_SECURE_ADDR, HWIO_APSS_CL_SECURE_RMSK);
+	mmio_write_32(HWIO_APSS_BANKED_APC_SECURE_ADDR, HWIO_APSS_BANKED_APC_SECURE_RMSK);
+	mmio_write_32(HWIO_APSS_SHARED_SHR_SECURE_ADDR, 0x0);
+#ifdef HWIO_APSS_OSM_TOP_OSM_SECURE_ACCESS_OVERRIDE_ADDR
```
Ditto, can we drop changes under this macro?

ldts replied:
yep

b49020 commented on drivers/qti/sec_core/sec_core.c:
```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
```
Add a copyright year since the TF-A CI checks require it.

ldts replied:
um,  checpatch didnt complain. ok

b49020 replied:
Yeah the TF-A CI has a lot more static checks additional to checkpatch which will complain.

b49020 commented on plat/qti/kodiak/inc/msm_hwio_base.h:
Let's not add these huge headers upstream. We should only add needed offset from the base address in the driver specific header where the base address comes from `kodiak_def.h` which will help with portability to other platforms.

b49020 commented on plat/qti/kodiak/inc/sec_core_def.h:
Ditto here, lets add macros which are really used to make the driver portable.

ldts replied:
lets see how much I Can trim...the driver uses more than it seems

b49020 commented on plat/qti/kodiak/rb3gen2/platform.mk:
```diff
@@ -97,6 +97,9 @@ BL31_SOURCES		+=	drivers/delay_timer/generic_delay_timer.c		\
 QTISECLIB_PATH ?=
 
+# QTISECLIB drivers
+BL31_SOURCES	+=			drivers/qti/sec_core/sec_core.c
```
Let's append this driver to the list above since it will be a proper TF-A driver.

ldts replied:
going to keep them here while the port is on-going.

b49020 replied:
That's fine with me.

b49020 commented on plat/qti/common/src/qti_pm.c:
```diff
@@ -286,6 +286,8 @@ int plat_setup_psci_ops(uintptr_t sec_entrypoint,
 {
 	int err;
 
+	qti_sec_core_remap((uintptr_t)bl31_warm_entrypoint);
```
Can we move this invocation after `qtiseclib_psci_init()` such that configuration done by `qti_sec_core_remap()` gets used? I suppose here we are setting up the entry point for secondary cores.

ldts replied:
I modified qtiseclib and removed the entry - is the only way I can reliably test.

b49020 replied:
Yeah I agree but for upstream we need to use the released blob only. It's better we override the configuration done by the blob instead and get it tested.

b49020 commented on drivers/qti/sec_core/sec_core.c:
```diff
@@ -0,0 +1,80 @@
+
+void qti_sec_core_remap(uintptr_t entrypoint)
+{
+	mmio_write_32(HWIO_APSS_SHARED_KRYO_RVBARADDR_LO_ADDR,
```
I suppose `HWIO_APSS_SHARED_KRYO_RVBARADDR_LO_ADDR` is a platform specific address which should come from `kodiak_def.h`.

ldts replied:
I am adding a kodiak_hwio.h which might end up being merged into kodiak_def.h.

ldts replied:
let me try to merge in one go (was wrapping up the qtimer and is similar in a way)

ldts committed:
feat(qti/kodiak): add sec_core driver

Sets the CPU reset start address (via a helper) and configures the
security/access policy for CPU/clock/power/timer blocks and the
interconnect so the rest of the boot flow and the non‑secure OS can
safely access what they need.

When this driver is enabled the Qualcomm propietary security library
(QTISECLIB) no longer needs to perform these actions.

Change-Id: I3fd6c62ce6b69f5d1e43f4207142ceaa07b9b370
Signed-off-by: Jorge Ramirez-Ortiz <jorge.ramirez@oss.qualcomm.com>

ldts committed:
feat(qti/kodiak): add qtimer driver

Introduce a minimal QTimer driver implementing basic initialization, raw
counter access, and a microsecond conversion helper.

Change-Id: I8dbde26848642b697302508847a7812e32d146e0
Signed-off-by: Jorge Ramirez-Ortiz <jorge.ramirez@oss.qualcomm.com>

ldts committed:
feat(qti): integrate qtimer stub

Add the qtimer driver stub and hook it into the generic QTI driver
framework. The stub remains disabled while QTISECLIB continues to
provide the active implementation.

For coverity purposes, enable the driver on the non-bootable build for
the kodiak/rb3gen2 platform.

Change-Id: Icd93ca442b473a7c38bbbb203d820497cf3d1ed3
Signed-off-by: Jorge Ramirez-Ortiz <jorge.ramirez@oss.qualcomm.com>

b49020 commented on plat/qti/kodiak/inc/qti_sec_core.h:
Can most of the macros in this header be moved to: `include/drivers/qti/sec_core/sec_core.h` to allow reuse?

b49020 commented on plat/qti/kodiak/inc/qti_sec_core.h:
```diff
@@ -0,0 +1,66 @@
+#ifndef QTI_SEC_CORE_H
+#define QTI_SEC_CORE_H
+
+#include "kodiak_def.h"
```
This can rather be `platform_def.h` which automatically includes `kodiak_def.h`. See `include/drivers/qti/crypto/rng.h` for reference.

b49020 commented on plat/qti/kodiak/inc/qti_qtimer.h:
```diff
@@ -0,0 +1,15 @@
+#ifndef QTI_QTIMER_H
+#define QTI_QTIMER_H
+
+#define QTIMER_BASE_ADDRESS		0x17C20000
```
This should move to `kodiak_def.h` and rest can move to `include/drivers/qti/qtimer/qtimer.h` such that we don't need any additional header.

b49020 commented on plat/qti/common/src/qti_pm.c:
```diff
@@ -10,9 +10,9 @@
+#include <drivers/qti/sec_core/sec_core.h>
 #include <lib/mmio.h>
 #include <lib/psci/psci.h>
-
```
This empty blank line between different types of header includes is rather enforced by TF-A static checks, so don't drop it.

b49020 commented on drivers/qti/qtimer/qtimer.c:
```diff
@@ -0,0 +1,48 @@
+#define VOFF_FG0_LO_OFFSET(n)		(0x80 + 0x8 * (n))
+#define VOFF_FG0_HI_OFFSET(n)		(0x84 + 0x8 * (n))
+#define ACR_FG0_OFFSET(n)		(0x40 + 0x4 * (n))
+#define SAR_FG0_OFFSET			(0x4)
```
I think these macros can rather move to `include/drivers/qti/qtimer/qtimer.h`.

b49020 reviewed (commented):
For commit description, don't include `/` and could rather be:

`feat(rb3gen2): enable sec_core driver`

ldts committed:
feat(qti/kodiak): add watchdog driver

Add a Kodiak platform watchdog driver providing timeout programming,
bark/bite handling, and per-core tracking, integrating with QTimer and
the generic interrupt service.

Change-Id: Ieb8cd5a597253553ac7f0f1b861cb78bff35a12d
Signed-off-by: Jorge Ramirez-Ortiz <jorge.ramirez@oss.qualcomm.com>

ldts committed:
feat(qti): integrate watchdog stub

Add the watchdog driver stub and hook it into the generic QTI driver
framework. The stub remains disabled while QTISECLIB continues to
provide the active implementation.

For coverity purposes, enable the driver on the non-bootable builds for
the kodiak/rb3gen2 platform.

Change-Id: I080bf836fb4de12b2f1b4dd1971b64220835079a
Signed-off-by: Jorge Ramirez-Ortiz <jorge.ramirez@oss.qualcomm.com>

ldts committed:
feat(qti): extend interrupt dispatcher

Expand the interrupt dispatcher so that device interrupts can be routed
to the new in-tree drivers instead of relying on QTISECLIB.

Change-Id: Ic23702ce27eaf18d6d01b151aabc18fb8b02196d
Signed-off-by: Jorge Ramirez-Ortiz <jorge.ramirez@oss.qualcomm.com>

ldts committed:
feat(qti/kodiak): add access control driver

Add the QTI accesscontrol driver, which wraps the memory-assignment API
and initializes the VMIDMT and XPU subsystems (static config plus
interrupt registration, with fatal-error handling on failure).

To use this implementation, the access control support provided by
QTISECLIB must be disabled at build time. When the driver is not
enabled, the associated stub continues to call into QTISECLIB.

Change-Id: Ib5472ea2a3b41b75faa4c8766bc9f0ad3d23061a
Signed-off-by: Jorge Ramirez-Ortiz <jorge.ramirez@oss.qualcomm.com>

ldts committed:
feat(qti): integrate access control stub

Enable the access control stub so platforms can transition away from
the QTISECLIB implementation. Using the real driver requires linking
against a QTISECLIB build where the corresponding functionality has
been disabled.

For coverity purposes, enable the driver on the non-bootable builds for
the kodiak/rb3gen2 platform.

Change-Id: I9f74a680b1bf758adf27e9bf6e04589cfce5305c
Signed-off-by: Jorge Ramirez-Ortiz <jorge.ramirez@oss.qualcomm.com>

ldts commented:
Please feel free to comment on functional behavior or additional abstractions - the actual review will happen upstream (dont be concerned about code formatting here)

b49020 commented:
@ldts 

> The user must disable the access control driver provided by the
QTISECLIB to be able to build this implementation.

Is it possible to have the access control driver override functionality provided by QTISECLIB blob publicly available here: `https://github.com/coreboot/qc_blobs/tree/master/sc7280/qtiseclib`? Or else we need to host changes for QTISECLIB internally for others to reproduce?

ldts commented:
> @ldts
> 
> > The user must disable the access control driver provided by the
> > QTISECLIB to be able to build this implementation.
> 
> Is it possible to have the access control driver override functionality provided by QTISECLIB blob publicly available here: `https://github.com/coreboot/qc_blobs/tree/master/sc7280/qtiseclib`? Or else we need to host changes for QTISECLIB internally for others to reproduce?

unfortunately, we can't since global symbols would conflict during build as those would also be present in the library. From the QTISECLIB perspective is pretty much just to remove a makefile so we can generate the updated library very simply.

`-# include    drivers/accesscontrol/qtisec.mk`

ldts commented:
@b49020 I'll propose this set upstream over the weekend - need to update the commit messages

ldts committed:
docs(maintainers): update QTI maintainers

Add myself to the list of QTI platform maintainers to include the
platform drivers.

Change-Id: I2746d74b182039a2db955e17313f8b439cfe6954
Signed-off-by: Jorge Ramirez-Ortiz <jorge.ramirez@oss.qualcomm.com>

ldts commented:
snapshot of what is in the upstream PR as today.

b49020 commented:
Boots fine on my RB3G2 board using publicly available QTISECLIB blob, merging now.
